src/node_sqlite.cc

```diff
@@ -506,6 +508,145 @@ void StatementSync::All(const FunctionCallbackInfo<Value>& args) {
+  LocalVector<Value> values(isolate,
+                            {Boolean::New(isolate, true), Null(isolate)});
+
+  Local<Object> result =
+      Object::New(isolate, Null(isolate), keys.data(), values.data(), 2);
+  args.GetReturnValue().Set(result);
+}
+
+void StatementSync::IterateNextCallback(
+    const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  auto isolate = env->isolate();
+
+  Local<External> data = Local<External>::Cast(args.Data());
+  IterateCaptureContext* capture_context =
+      static_cast<IterateCaptureContext*>(data->Value());
+  auto stmt = capture_context->stmt;
+  auto num_cols = capture_context->num_cols;
+
+  int r = sqlite3_step(stmt->statement_);
+  if (r != SQLITE_ROW) {
+    CHECK_ERROR_OR_THROW(
+        env->isolate(), stmt->db_->Connection(), r, SQLITE_DONE, void());
+    sqlite3_reset(stmt->statement_);
+
+    LocalVector<Name> keys(isolate,
+                           {String::NewFromUtf8Literal(isolate, "done"),
+                            String::NewFromUtf8Literal(isolate, "value")});
+    LocalVector<Value> values(isolate,
+                              {Boolean::New(isolate, true), Null(isolate)});
+
+    Local<Object> result =
+        Object::New(isolate, Null(isolate), keys.data(), values.data(), 2);
+    args.GetReturnValue().Set(result);
+    return;
+  }
+
+  LocalVector<Name> row_keys(isolate);
+  row_keys.reserve(num_cols);
+  LocalVector<Value> row_values(isolate);
+  row_values.reserve(num_cols);
+  for (int i = 0; i < num_cols; ++i) {
+    Local<Name> key;
+    if (!stmt->ColumnNameToName(i).ToLocal(&key)) return;
+    Local<Value> val;
+    if (!stmt->ColumnToValue(i).ToLocal(&val)) return;
+    row_keys.emplace_back(key);
+    row_values.emplace_back(val);
+  }
+
+  Local<Object> row = Object::New(
+      isolate, Null(isolate), row_keys.data(), row_values.data(), num_cols);
+
+  LocalVector<Name> keys(isolate,
+                         {String::NewFromUtf8Literal(isolate, "done"),
+                          String::NewFromUtf8Literal(isolate, "value")});
+  LocalVector<Value> values(isolate, {Boolean::New(isolate, false), row});
+
+  Local<Object> result =
+      Object::New(isolate, Null(isolate), keys.data(), values.data(), 2);
+  args.GetReturnValue().Set(result);
+}
+
+void StatementSync::Iterate(const FunctionCallbackInfo<Value>& args) {
+  StatementSync* stmt;
+  ASSIGN_OR_RETURN_UNWRAP(&stmt, args.This());
+  Environment* env = Environment::GetCurrent(args);
+  THROW_AND_RETURN_ON_BAD_STATE(
+      env, stmt->IsFinalized(), "statement has been finalized");
+  auto isolate = env->isolate();
+  auto context = env->context();
+  int r = sqlite3_reset(stmt->statement_);
+  CHECK_ERROR_OR_THROW(
+      env->isolate(), stmt->db_->Connection(), r, SQLITE_OK, void());
+
+  if (!stmt->BindParams(args)) {
+    return;
+  }
+
+  IterateCaptureContext* capture_context = new IterateCaptureContext();
```

Comment:
I don't know how to solve it, but I'm pretty sure there is a memory leak because this object is never destroyed.

Reply:
I supposed `External::New(isolate, capture_context))` manage destroy with garbage collector.

Reply:
Maybe, I don't know.

Reply:
No, I do not believe that `External` destroys this. You'll have to arrange for it to be destroyed when the external is collected.

Reply:
How can I do that ?

Reply:
I tried to do something on this memory leak: https://github.com/nodejs/node/pull/54213/commits/fb13c9fa6256a467803c3360b50c4a712d922686
But it is not working as expected. I'm not used to C/C++ and how to allocate and free memory. And I have no idea on how to do it with v8.
I need help on this topic please.

Reply:
@joyeecheung Maybe you can help? This is also too complicated for me.

Reply:
https://github.com/nodejs/node/pull/54213/commits/d89cb556d49ad8b96d7066cd7579e1f86f58b017

I used `.This` instead of `.Data` and use PropertyDescriptor to store data needed for `IterateReturnCallback` and `IterateNextCallback` (no `IterateCaptureContext*` needed anymore).
I hope this remove the memory leak.

Reply:
Like `All` method, `IterateNextCallback` and `IterateReturnCallback` reset the statement (I guess it free some memory before bind new params and reuse the statement).

To avoid potential edge case with manual call to the iterator, when the iteration is over, `is_finished` property is set to true. `IterateNextCallback` will early return without touching the statement.

Reply:
One week with no response, I guess it's good so I resolve.